File: backend/stageops/catalog.py
import json
from datetime import UTC, date, datetime
from importlib.resources import files
from uuid import uuid4
from zoneinfo import ZoneInfo

from sqlalchemy.dialects.postgresql import insert
from sqlalchemy.orm import Session

from .models import Asset, Product, Scenario

FIXTURE = json.loads(files("stageops").joinpath("data/catalog.json").read_text())


def first_of_month_after(day: date, months: int = 1) -> date:
    year, month = divmod(day.year * 12 + day.month - 1 + months, 12)
    return date(year, month + 1, 1)


def rental_window(now: datetime | None = None, time_zone: str = "Asia/Tashkent") -> dict:
    today = (now or datetime.now(UTC)).astimezone(ZoneInfo(time_zone)).date()
    return {"today": today, "minDate": first_of_month_after(today), "timeZone": time_zone}


def seed_catalog(session: Session) -> None:
    for position, product in enumerate(FIXTURE["products"]):
        values = {key: product[key] for key in ("id", "category", "name", "kind", "detail")}
        session.execute(
            insert(Product).values(**values, position=position).on_conflict_do_nothing()
        )


def create_scenario(session: Session, today: date) -> Scenario:
    month = first_of_month_after(today)
    scenario = Scenario(id=uuid4(), month=month, created_at=datetime.now(UTC))
    session.add(scenario)
    session.flush()
    later = first_of_month_after(month, 3)
    for product in FIXTURE["products"]:
        for index in range(product["total"]):
            if index < product["freeUnits"]:
                available = today
            elif index == product["freeUnits"]:
                available = month.replace(day=product["releaseDay"])
            else:
                available = later
            session.add(
                Asset(scenario_id=scenario.id, product_id=product["id"], available_from=available)
            )
    session.flush()
    return scenario

File: lib/catalog.ts
import fixture from "../backend/stageops/data/catalog.json" with { type: "json" };

export type CategoryId = "sound" | "signal" | "backline";
export type Product = {
  id: string;
  category: CategoryId;
  name: string;
  kind: string;
  detail: string;
};
export type SeedProduct = Product & { total: number; releaseDay: number; freeUnits: number };
export type Category = { id: CategoryId; title: string; short: string; description: string; image: string; total: number };
export const categories = fixture.categories as Category[];
export const products = fixture.products as SeedProduct[];
export type RentalRange = { start: string; end: string };
export type CartLine = {
  productId: string;
  quantity: number;
  range?: RentalRange;
};
export function isoDate(date: Date): string {
  return `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, "0")}-${String(date.getDate()).padStart(2, "0")}`;
}
export function parseDate(value: string): Date {
  const [year, month, day] = value.split("-").map(Number);
  return new Date(year, month - 1, day, 12);
}
export function shiftDay(value: string, days: number): string {
  const date = parseDate(value);
  date.setDate(date.getDate() + days);
  return isoDate(date);
}
export function rentalWindow(now = new Date()) {
  const today = new Intl.DateTimeFormat("en-CA", {
    timeZone: "Asia/Tashkent",
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  }).format(now);
  const date = parseDate(today);
  return {
    today,
    minDate: isoDate(new Date(date.getFullYear(), date.getMonth() + 1, 1)),
    timeZone: "Asia/Tashkent",
  };
}
export function shortDate(value: string) {
  return parseDate(value)
    .toLocaleDateString("ru-RU", { day: "numeric", month: "short" })
    .replace(".", "");
}
export function rangeLabel(range?: RentalRange) {
  return range
    ? `${shortDate(range.start)} — ${shortDate(range.end)}`
    : "Выберите даты";
}
